### ISSUE TITLE
Allow user to enable/disable bounds checking

### ISSUE THREAD

**Issue by kevinushey** [RcppCore/Rcpp#137]
Ie, allow the user to control this with a define macro, similar to Armadillo. Maybe turn it on by default and allow the user to define `RCPP_NO_DEBUG` to turn it off.

**Comment by coatless** [comment]
Not quite a 1-to-1 but, bounds checks were discussed in #341 and implemented by PR #342. 

In summary:

To sum up:
operator[]: no bounds check, unchanged
operator(): no bounds check
operator at(): bounds check was added

Perhaps close? (Tagging #506)

**Comment by coatless** [comment]
I think this can be closed out since the addition of a `#define` to disable bounds check globally was rejected.

**Comment by coatless** [comment]
Past @coatless, what were you thinking requesting this be closed? A quarter of the new StackOverflow posts have been directly related to a lack of bounds checks due to examples showing code that negates the bounds check!

So, what I think would be helpful:

- add a `#define` to disable the bounds in a manner similar to [armadillo's `#define ARMA_NO_DEBUG`](http://arma.sourceforge.net/docs.html#element_access)... Perhaps, `#define RCPP_NO_BOUNDS_CHECK` ?
- add an FAQ entry on bounds checking
- convert existing documentation entries to use bounds checked code.

Thoughts? @eddelbuettel

**Comment by eddelbuettel** [comment]
```r
> Rcpp::cppFunction("bool foo(NumericVector x) { return x[5] > 0; }")
> x <- 1:10; foo(x)
[1] TRUE
> 
> x <- 1:3; foo(x)
[1] TRUE
> 
> Rcpp::cppFunction("bool bar(NumericVector x) { return x(5) > 0; }")
> x <- 1:3; bar(x)
Error in bar(x) : Index out of bounds: [index=5; extent=3].
> 
```

For vectors, we have `x(i)` vs `x[i]`.  I still think that may be good enough.

**Comment by coatless** [comment]
Well, bounds checks exist for both vectors, `()`, and matrices,`.at()`.  The issue is not a lot of folks are using the bounds checks by default and are running into UBs as a result of OOBs (c.f. stackoverflow posts in the past month).

The goal here would be to add a `#define` (think: `#define RCPP_NO_BOUNDS_CHECKS`) that allows the bounds check to be disabled on `()` and `.at()` **after the routine has been thoroughly tested.** In essence, the `#define` would transform accessor `()` and `.at()` to `[]` and `()` respectively. Thus, users get the benefit of bounds checking while developing code and the ability to maximize its speed by simply removing the bounds check.

**Comment by eddelbuettel** [comment]
I understand.

But I take your logic of "sure it exists but poor lusers don't find" and apply it to a random `#define` they are also unable to utlize.

News at 11pm.

**Comment by coatless** [comment]
@eddelbuettel True. But, the status quo now basically mandates users switch each instances of the accessors back and forth depending if they want bounds checking or not. Adding the optional `#define` to disable bounds checking on `()` for vectors and `.at()` for matrices would be nice. I take advantage of this feature often when working with Armadillo. 

Regardless, these behaviors should at least warrant an entry in the FAQ.

**Comment by coatless** [comment]
Another one... 

https://stackoverflow.com/questions/47885216/memory-leaks-in-a-simple-rcpp-function

I'm thinking about adding the ability to disable bounds checks by setting `RCPP_DEBUG_LEVEL` define to -1. Presently, this macro is initialized at 0. 

https://github.com/RcppCore/Rcpp/blob/1630a55be532ea35bb4a9e1d4afba7d81aa32981/inst/include/Rcpp/macros/debug.h#L25-L28

This would allow the bounds check to be disabled without adding yet another macro definition.

Thoughts?

**Comment by eddelbuettel** [comment]
I saw the SO post. 

To me it perfectily well demonstrates that we already can have bounds checking, and I see no need or desire to further complicating code with a new `#ifdef` branch at every access.  Other views?

**Comment by coatless** [comment]
Pinging @kevinushey @nathan-russell

**Comment by kevinushey** [comment]
I'm a bit confused. We're talking about disabling (?) bounds checking, but https://stackoverflow.com/questions/47885216/memory-leaks-in-a-simple-rcpp-function is an instance where the user was _not_ saved by bounds checking (instead, the program just crashed later after an out-of-bounds memory write).

Since the user was using `operator[]` here, no bounds checking was performed. We decided a while back that this is how it should be (for performance reasons).

I don't think `RCPP_DEBUG_LEVEL` is the right macro to hang this off of, as setting that causes Rcpp to spew out a _lot_ of diagnostic information. It's primarily for Rcpp developers, to see what methods are actually being invoked in a bit of Rcpp code.

**Comment by eddelbuettel** [comment]
Right. I also do not think that a user would be help by a new `#define` because that type of user typically does NOT recompile their Rcpp installation.  And valgrind/ASAN/UBSAN/... all get these things too these days.  So to me, the status quo in Rcpp is fine.

**Comment by coatless** [comment]
@kevinushey

The idea here is to mimic [armadillo's bounds check initiative](http://arma.sourceforge.net/docs.html#element_access). In general, users are _encouraged_ to write code that uses the bounds check mechanism and once their code is debugged to switch off the bounds check during compile-time to gain more speed. 

For example, we have:

```
#include<armadillo.h>
using namespace arma;

void sample_access() { 
  // Bounds checking access
  mat A = randu<mat>(10,10);
  A(9, 9) = 123.0;
  double y = A(99);
}
```

To:

```
// Disabled bounds checking access
#define  ARMA_NO_DEBUG 1
#include<armadillo.h>
using namespace arma;

void sample_access() { 
  // Traditional bounds check is no longer active
  mat A = randu<mat>(10,10);
  A(9, 9) = 123.0;
  double y = A(99);
}
```

I believe having this option and _emphasizing_ the bounds check operator within documentation would lower the instance rate of the off-by-one errors. I'm making this suggestion as this I've likely dealt with ~10 questions on SO and more working with students due to translation issues resulting in _R_'s system using 1-based indices vs. _C++_'s 0-base indices. 

In conclusion, the proposed change to _Rcpp_ would be two fold:

1. Modify the bounds check operators, e.g. `(i)` and `.at(i,j)`, to suppress the check if a `#define` was declared. 
2.  Update documentation to show `(i)` and `.at(i,j)` operators over `[i]` and `(i,j)`

**Comment by eddelbuettel** [comment]
a) Well, as it exists in Armadillo, we can (and have !!) always tell people to use (Rcpp)Armadillo.

b) If you must, you could always fork or branch Rcpp so that you can put as many bells and whistles and documentation enhancements in as you please.

c) I find our code complex enough as it stands and see no gain from mucking with _all_ element accessors for no apparent reason other than to _potentially_ help newbies -- who still would *not* turn the required mechanism on.

So in sum: too much ado for essentially nothing.  Still not a fan, and would like to put this to rest for good.

**Comment by eddelbuettel** [comment]
> Regardless, these behaviors should at least warrant an entry in the FAQ.

That still hold, and is the best place.  We can even help with a Gallery post on valgrind or ASAN.  

(More for all of us:  [Jason Turner on Sanitizers](https://www.youtube.com/watch?v=MB6NPkB4YVs))
